lib/Parser/Parser.cpp

```diff
@@ -641,36 +623,32 @@ Optional<AST> Parser::parseAST() {
     Type t;
     Effect e;
 
-    bool changed;
-    do {
-        changed = false;
-        if(parsePredicate().unwrapInto(p)) {
-            predicates.push_back(p);
-            changed = true;
-        }
-
-        if(parseType().unwrapInto(t)) {
-            types.push_back(t);
-            changed = true;
-        }
+    try {
+        bool reached_eof = false;
 
-        if(parseEffect().unwrapInto(e)) {
-            effects.push_back(e);
-            changed = true;
-        }
-    } while(changed);
+        do {
+            if(parsePredicate().unwrapInto(p)) {
+                predicates.push_back(p);
+            } else if(parseType().unwrapInto(t)) {
+                types.push_back(t);
+            } else if(parseEffect().unwrapInto(e)) {
+                effects.push_back(e);
+            } else {
+                Token unexpectedToken = lexer.peek_next();
+                if(unexpectedToken.type == Token::Type::end_of_file) {
+                    reached_eof = true;
+                } else {
+                    throw(SyntaxError("Unexpected token \"" + unexpectedToken.text + "\"."));
+                }
+            }
+        } while(!reached_eof);
 
-    if(foundSyntaxError) {
-        return Optional<AST>();
-    } else {
         return AST(types, effects, predicates);
+    } catch (SyntaxError& se) {
```

Comment:
I ended up going for the approach of ignoring error recovery completely and only allowing one syntax error per execution. The easiest way I could think of to terminate immediately when we run into a syntax error was to throw a SyntaxError and catch it in the top-level parsing method. All locations where `emitSyntaxError()` used to be called have been replaced with `throw(SyntaxError("message"))`, and the message generation logic that used to be in `emitSyntaxError()` has been moved to the `catch()` block here. Let me know if you're okay with this implementation.